File: haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation (proteins) where

import Data.List.Split (chunksOf)

proteins :: String -> Maybe [String]
proteins = proteins' . chunksOf 3
  where
    proteins' [] = Just []
    proteins' (x : xs) = case lookup x codonToProtein of
      Nothing -> Nothing
      Just protein -> if protein == "STOP" then Just [] else (protein :) <$> proteins' xs

codonToProtein :: [(String, String)]
codonToProtein =
  [ ("AUG", "Methionine"),
    ("UUU", "Phenylalanine"),
    ("UUC", "Phenylalanine"),
    ("UUA", "Leucine"),
    ("UUG", "Leucine"),
    ("UCU", "Serine"),
    ("UCC", "Serine"),
    ("UCA", "Serine"),
    ("UCG", "Serine"),
    ("UAU", "Tyrosine"),
    ("UAC", "Tyrosine"),
    ("UGU", "Cysteine"),
    ("UGC", "Cysteine"),
    ("UGG", "Tryptophan"),
    ("UAA", "STOP"),
    ("UAG", "STOP"),
    ("UGA", "STOP")
  ]

File: typescript/protein-translation/protein-translation.ts
type Codon = string
type Protein = string

export const translate = (rna: string): string[] => {
  const transform = (codon: Codon[]): Protein[] => {
    if (codon.length === 0) return []

    const protein = asProtein(codon[0])
    if (protein === 'STOP') return []

    return [protein, ...transform(codon.slice(1))]
  }

  return transform(codons(rna))
}

const codons = (rna: string): Codon[] => {
  return rna.match(/.{1,3}/g) ?? []
}

const asProtein = (codon: Codon): Protein => {
  if (codon === 'AUG') return 'Methionine'
  if (codon === 'UUU' || codon === 'UUC') return 'Phenylalanine'
  if (codon === 'UUA' || codon === 'UUG') return 'Leucine'
  if (codon === 'UCU' || codon === 'UCC' || codon === 'UCA' || codon === 'UCG') return 'Serine'
  if (codon === 'UAU' || codon === 'UAC') return 'Tyrosine'
  if (codon === 'UGU' || codon === 'UGC') return 'Cysteine'
  if (codon === 'UGG') return 'Tryptophan'
  if (codon === 'UAA' || codon === 'UAG' || codon === 'UGA') return 'STOP'

  console.error(`Invalid codon: ${codon}`)
  throw new Error('Invalid codon')
}

